### DISCUSSION TITLE
Product Image block: Image is rendering with different size on the frontend

### DISCUSSION THREAD

**Discussion by thealexandrelara** [woocommerce/woocommerce#44971] [WooCommerce, Blocks, and Site Editing]
Recently, I've been working on the development of the Single Product block. This block comprises a set of inner blocks, such as Product Image, Product Title, Product Summary, Product Rating, etc.

While reviewing the implementation, @Aljullu [found](https://github.com/woocommerce/woocommerce-blocks/pull/8610#pullrequestreview-1365216995) that when inside the Editor the Product Image appeared at its full size, but when looking on the frontend the image was being displayed cropped (as a thumbnail):

Editor | Frontend
--- | ---
![image](https://user-images.githubusercontent.com/3616980/228861815-0ea0dccd-de21-4993-9066-2376957974c3.png) | ![image](https://user-images.githubusercontent.com/3616980/228862037-9aefef58-2b51-4739-be57-c0df629075d5.png)

## Why?
As stated by @Aljullu this was probably occurring "because in the frontend we are rendering a thumbnail and in the editor the original (or something like that)". So, taking that as a premise, I opened an issue to investigate it further, and my conclusion is that this is precisely what is happening.

Looking at the ProductImage class (on src/BlockTypes/ProductImage.php), we have the following code to render the image on the frontend:

```js
private function render_image( $product ) {
	$image_info = wp_get_attachment_image_src( get_post_thumbnail_id( $product->get_id() ), 'woocommerce_thumbnail' );

	if ( ! isset( $image_info[0] ) ) {
		return sprintf( '<img src="%s" alt="" />', wc_placeholder_img_src( 'woocommerce_thumbnail' ) );
	}

	return sprintf(
		'<img data-testid="product-image" alt="%s" src="%s">',
		$product->get_title(),
		$image_info[0]
	);
}
```
WooCommerce provides [different options for the image sizes](https://woocommerce.com/document/image-sizes-theme-developers/): `woocommerce_thumbnail`, `woocommerce_single`, and `woocommerce_gallery_thumbnail`; however, as we can see in the above code we are always rendering the thumbnail image by using the hardcoded option `woocommerce_thumbnail`. 

I noticed that the Product Image has [the attribute `imageSizing` that defaults to `full-size`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/atomic/blocks/product-elements/image/attributes.ts#L19-L22), but the attribute is being ignored when rendering the image (as in the code shared above)

### Proposal
I propose that we should start considering the `imageSizing` attribute inside the `render_image`:
```
private function render_image( $product, $attributes ) {
	$image_type = 'full-size' == $attributes['imageSizing']  ? 'woocommerce_single' : 'woocommerce_thumbnail';
	$image_info = wp_get_attachment_image_src( get_post_thumbnail_id( $product->get_id() ), $image_type );

	if ( ! isset( $image_info[0] ) ) {
		return sprintf( '<img src="%s" alt="" />', wc_placeholder_img_src( $image_type ) );
	}

	return sprintf(
		'<img data-testid="product-image" alt="%s" src="%s">',
		$product->get_title(),
		$image_info[0]
	);
}
```
With this approach we would be able to render the image depending on the value set on the `imageSizing` attribute.

### The issue
I've started mapping some of the blocks that makes use of the Product Image block:
```
Related Products
Products (Beta)
Single Product
Cart / Checkout (not sure if these are actually using the Product Image block)
```

While testing this approach in different blocks I noticed that if we render the `woocommerce_single` when the imageSizing attribute is set to `full-size` it has an impact in some blocks. For instance, the Product (Beta) block will render the full size image inside the images grid, so the product images that does not have an squared aspect ratio will be displayed in a different way:
<img width="896" alt="image" src="https://user-images.githubusercontent.com/20469356/229613359-841819d3-0d10-46b1-a795-78b34b60ec30.png">

### Solution 1 - Change the imageSizing attribute default to be `cropped`
Since the current image being displayed on the frontend is the "cropped" (thumbnail) version, maybe we should embrace that and defaults the `imageSizing` attribute to be `cropped`, so for the blocks that need to render the full size image (like the Single Product block) we can set this attribute to be `full-size`.

### Solution 2 - Set the `imageSizing` attribute on each block
Instead of changing the default value, we can just set the `imageSizing` for each block, instead of relying on its default behavior. The idea would be to map the blocks using the Product Image block and define whether we want to display the image at its full size.

## Questions
1. Do we have any other block that this change could impact?
2. Which solution would be preferable? Do you see any other solution to solve this issue?
3. It looks like the `imageSizing` attribute cannot be changed/configured by the merchants. Should we allow them to modify this property on the Editor?
4. Should we keep using `full-size` and `cropped` as values for the `imageSizing` attribute, or would it be better to modify it to be the same as WooCommerce: `thumbnail`, `single`, and `gallery_thumbnail`?

**Comment by Aljullu** [comment]
Thanks for investigating this, @thealexandrelara!

> I propose that we should start considering the `imageSizing` attribute inside the `render_image`:

Completely agree. :100: 

> Do we have any other block that this change could impact?

We have some other blocks that display product images, but AFAIK they don't use the Product Image inner block, so they won't be affected. Btw, from your list, I think Cart and Checkout don't use the Product Image inner block either.

> Which solution would be preferable? Do you see any other solution to solve this issue?

I don't have a strong opinion, but I would go with:

1. Product Image has `full-size` by default.
2. We override it in the Related Products and Products blocks to be `cropped`, so all product images show the same size in the grid.

> It looks like the `imageSizing` attribute cannot be changed/configured by the merchants. Should we allow them to modify this property on the Editor?

You're right. This feature is currently only available in the Customizer, which is not available in block themes. There is some more context here: pdnLyh-181-p2. I wonder if we should add a link to the Customizer from the Product Image block. :thinking: 

> Should we keep using `full-size` and `cropped` as values for the `imageSizing` attribute, or would it be better to modify it to be the same as WooCommerce: `thumbnail`, `single`, and `gallery_thumbnail`?

Good point. I don't have a strong opinion either, so I'm fine either way.

**Reply by thealexandrelara** [reply]
Thank you @Aljullu for your contribution to this discussion! 

> I don't have a strong opinion, but I would go with:
>
> Product Image has full-size by default.
> We override it in the Related Products and Products blocks to be cropped, so all product images show the same size in the grid.

I agree with this, we can keep `full-size` as being the default and set the attribute for each block.

> You're right. This feature is currently only available in the Customizer, which is not available in block themes. There is some more context here: pdnLyh-181-p2. I wonder if we should add a link to the Customizer from the Product Image block. 🤔

I think it would be good to have the link showing there, but is the Customizer available for block themes?

> Good point. I don't have a strong opinion either, so I'm fine either way.

Nice! I'll investigate how hard it would be to change, but if it is a trivial thing maybe following the same naming as WooCommerce could be beneficial and avoid any confusion in the future or the need to map values like:
```
cropped = thumbnail
full-size = single
cropped-gallery = gallery_thumbnail
...
```

**Reply by Aljullu** [reply]
> I think it would be good to have the link showing there, but is the Customizer available for block themes?

The thing is that the _Customizer_ is hidden from the UI, but it can be accessed directly with the URL. But yeah, I agree with you, let's not add the link for now.

> Nice! I'll investigate how hard it would be to change, but if it is a trivial thing maybe following the same naming as WooCommerce could be beneficial and avoid any confusion in the future or the need to map values like:

:+1:

**Reply by gigitux** [reply]
Thanks for the exploration, @thealexandrelara! I agree with this approach!

If we decide to use `full-size` as default, we have to pay attention to not break the UI of the current instances of the Product Image block used by the Products block.

The new attribute will be available for the new Product Image blocks that will be added but will not exist for the old one.

**Comment by akshaynathani** [comment]
[https://pasteboard.co/Og0WMUtddkmv.jpg](url)

i have a similar issue. could you please help me to fix that ?
![Capture](https://github.com/woocommerce/woocommerce/assets/66509235/4169e46a-4160-4e91-994c-4a1e2b762f6b)

**Reply by akshaynathani** [reply]
as you can see the image is not displaying properly, what are the ways to fix it ?
